feat: isolate context by different appoach

MaikoTan opened the pull request:
Per https://github.com/koishijs/chatgpt-bot/issues/8#issuecomment-1340258845

So then we can maintain the conversation context history within one channel.

为了 https://github.com/koishijs/chatgpt-bot/issues/8#issuecomment-1340258845

于是我们可以在一个频道里维持对话上下文历史。

-------

cc @shigma @TimeBather

MaikoTan committed:
feat: isolate context by different appoach

MaikoTan committed:
fix: union accepts const

MaikoTan commented on src/index.ts:
```diff
@@ -16,6 +20,11 @@ export const Config: Schema<Config> = Schema.intersect([
       Schema.array(String),
       Schema.transform(String, (prefix) => [prefix]),
     ] as const).description('使用特定前缀触发对话。').default(['!', '！']),
+    conversationContext: Schema.union([
+      Schema.const('user' as const).description('用户独立'),
+      Schema.const('channel' as const).description('频道独立'),
+      Schema.const('both' as const).description('频道内用户独立'),
+    ]).description('对话历史上下文缓存方式。').default('channel'),
```
Are there better naming and/or better translations?

TimeBather replied:
The "conversationContextIsolation" may better than "conversationContext"
And for the description , i think it has a better translation:"对话上下文隔离方式"

MaikoTan replied:
That's too loooooong.

I'd like a shorter one.

TimeBather replied:
How about "contextIsolation" .It's shorter than the second one.

MaikoTan replied:
Well, in my opinion, it would be better to focus on "sharing a context" rather than "isolate a context". So my suggestion is "contextInteraction".

TimeBather replied:
It's looks ok

MaikoTan replied:
Can you provide a translation of it in Chinese as well?

TimeBather replied:
I think it's a good choice to use "上下文共享模式(or 方式)"
If you think that it would be better to focus on "sharing a context" rather than "isolate a context" , it's better to change the translation of options from '独立' to '共享' . It's even better to change it from "频道独立" to "频道内独立"

Update:Fix some typo

MaikoTan replied:
```suggestion
    conversationContext: Schema.union([
      Schema.const('user' as const).description('用户共享'),
      Schema.const('channel' as const).description('频道共享'),
      Schema.const('both' as const).description('频道内用户独立'),
    ]).description('上下文共享方式。').default('channel'),
```

See if this satisfy yours?

shigma replied:
I think we can simply use `interaction` since there is no ambiguity.

MaikoTan replied:
Well, tasikani. Since we have description for every configuration items, it should be good.

shigma replied:
上下文共享方式：

- 用户上下文
- 频道上下文
- 频道+用户上下文

要不要这样改？

MaikoTan replied:
I think the first two phases make sense, but the last one might baffles people?

MaikoTan commented on src/index.ts:
```diff
@@ -26,6 +35,17 @@ export function apply(ctx: Context, config: Config) {
+        return (session: Session) => session.uid
+      case 'channel':
+        return (session: Session) => session.cid
+      case 'both':
+        return (session: Session) => `${session.cid}:${session.uid}`
```
Is it correct to diffre users with this?

shigma replied:
`${platform}:${channelId}:${userId}`

TimeBather commented:
And for the "getMapKey" function , I think there is a better naming , "getContextRecoverKey"  or "getContextStorageKey" , which means "Get the key for recover/storage the Context"

MaikoTan committed:
refactor: better naming

shigma commented on src/index.ts:
```diff
@@ -1,11 +1,15 @@
 const logger = new Logger('chatgpt')
 
+const interaction = ['user', 'channel', 'both'] as const
+export type ContextInteraction = typeof interaction[number]
```
```suggestion
export type Interaction = typeof interaction[number]
```

shigma commented on src/index.ts:
```diff
@@ -1,11 +1,15 @@
 export interface Config extends ChatGPT.Config {
   appellation: boolean
   prefix: string[]
+  interaction: ContextInteraction
```
```suggestion
  interaction: Interaction
```

shigma commented on src/index.ts:
```diff
@@ -16,6 +20,11 @@ export const Config: Schema<Config> = Schema.intersect([
+      Schema.const('user' as const).description('用户独立'),
+      Schema.const('channel' as const).description('频道独立'),
+      Schema.const('both' as const).description('频道内用户独立'),
+    ]).description('对话历史上下文缓存方式。').default('channel'),
```
总觉得现在的说法更抽象更难懂了……

MaikoTan replied:
see if the current one satisfies you, I do even add a document for it lol.

shigma commented on src/index.ts:
```diff
@@ -26,6 +35,17 @@ export function apply(ctx: Context, config: Config) {
 
   const api = new ChatGPT(ctx, config)
 
+  const getContextResolver = (() => {
```
其实我建议直接把 config 和 session 作为两个参数，这样更 reactive。

MaikoTan committed:
i18n: update translation of interaction option

MaikoTan committed:
refactor: better function naming

MaikoTan committed:
refactor: rename to Interaction

MaikoTan committed:
docs: add a comprehensive comment

MaikoTan committed:
docs: add document for conversation context

MaikoTan commented:
I need people to review this Pull Request once again, especially the document for the conversation context. Then this Pull Request can be merged as soon as possible!

我需要人们再次审查这个拉取请求，尤其是对话上下文的文档。那么这个拉取请求可以尽快合并！

shigma committed:
refactor: adjust naming

- move jsdoc comments
- use key instead of resolver - the latter sounds like a callback
